picoquic/ptls_openssl_backend.c

```diff
@@ -0,0 +1,23 @@
+/*
+* Copyright (c) 2020, Private Octopus, Inc., The PicoQuic Authors.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/**
+ * PicoTLS with openssl backend [WIP]
+ */
```

Comment:
Fusion is interesting from an API design point of view. The code in Picotls only runs on x64 platforms that support the AES instruction set. The application could try to use #ifdef to only add if fusion if supported, but that's error prone. Picotls has a function "int ptls_fusion_is_supported_by_cpu(void)" that can be called at run-time, and that's much more reliable.

In practice, that function will have to run during the "setup" call to the API. If the function returns False, the AESGCM suites will have to be provided differently -- minicrypto, openssl, maybe bcrypt on Windows.

There are two choices for the crypto-API: either have setup return an error, in which case the application will have to try a different provider; or require the "fusion" provider to do the fallback internally. I think letting the application decide is better.